src/pytorch_lightning/loggers/neptune.py

```diff
@@ -602,8 +603,8 @@ def _get_full_model_name(model_path: str, checkpoint_callback: "ReferenceType[Ch
     @classmethod
     def _get_full_model_names_from_exp_structure(cls, exp_structure: dict, namespace: str) -> Set[str]:
         """Returns all paths to properties which were already logged in `namespace`"""
-        structure_keys = namespace.split(cls.LOGGER_JOIN_CHAR)
-        uploaded_models_dict = reduce(lambda d, k: d[k], [exp_structure, *structure_keys])
+        structure_keys: list = namespace.split(cls.LOGGER_JOIN_CHAR)
```

Comment:
```suggestion
        structure_keys: List[str] = namespace.split(cls.LOGGER_JOIN_CHAR)
```

Reply:
when I make this change locally, the following error occurs:

```sh
src/pytorch_lightning/loggers/neptune.py:607: error: List item 1 has incompatible type "List[str]"; expected "Dict[Any, Any]"  [list-item]
```

Given the above, I updated ln 607 to 

```python
uploaded_models_dict: Dict[str, List[str]] = reduce(lambda d, k: d[k], [exp_structure, *structure_keys])
```

however, mypy then complained:

```sh
src/pytorch_lightning/loggers/neptune.py:607: error: Incompatible return value type (got "List[str]", expected "Dict[str, List[str]]")  [return-value]
src/pytorch_lightning/loggers/neptune.py:607: error: Invalid index type "Dict[str, List[str]]" for "Dict[str, List[str]]"; expected type "str"  [index]
src/pytorch_lightning/loggers/neptune.py:607: error: List item 1 has incompatible type "List[str]"; expected "Dict[str, List[str]]"  [list-item]
```

Likely because the lambda in reduce is using recursion to create the Dict `uploaded_models_dict`.

No errors occur when I leave the code as:

```python
def _get_full_model_names_from_exp_structure(cls, exp_structure: dict, namespace: str) -> Set[str]:
    """Returns all paths to properties which were already logged in `namespace`"""
    structure_keys: List = namespace.split(cls.LOGGER_JOIN_CHAR)
    uploaded_models_dict: Dict = reduce(lambda d, k: d[k], [exp_structure, *structure_keys])
    return set(cls._dict_paths(uploaded_models_dict))
```

Reply:
The reduction there is not compatible with the typing, so I changed it to a for loop. I'm not sure if it's "better" but this is the only way I found to make the types at least somewhat specific. Note, to properly type the dict there we would need some sort of recursive type definition, and I think that's not possible at the moment.